sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -62,20 +61,109 @@ case class CustomShuffleReaderExec private(
     }
   }
 
-  override def stringArgs: Iterator[Any] = Iterator(description)
+  override def stringArgs: Iterator[Any] = {
+    val desc = if (isLocalReader) {
+      "local"
+    } else if (hasCoalescedPartition && hasSkewedPartition) {
+      "coalesced and skewed"
+    } else if (hasCoalescedPartition) {
+      "coalesced"
+    } else if (hasSkewedPartition) {
+      "skewed"
+    } else {
+      ""
+    }
+    Iterator(desc)
+  }
 
-  private var cachedShuffleRDD: RDD[InternalRow] = null
+  def hasCoalescedPartition: Boolean = {
+    partitionSpecs.exists(_.isInstanceOf[CoalescedPartitionSpec])
+  }
 
-  override protected def doExecute(): RDD[InternalRow] = {
-    if (cachedShuffleRDD == null) {
-      cachedShuffleRDD = child match {
-        case stage: ShuffleQueryStageExec =>
-          new ShuffledRowRDD(
-            stage.shuffle.shuffleDependency, stage.shuffle.readMetrics, partitionSpecs.toArray)
-        case _ =>
-          throw new IllegalStateException("operating on canonicalization plan")
+  def hasSkewedPartition: Boolean = {
+    partitionSpecs.exists(_.isInstanceOf[PartialReducerPartitionSpec])
+  }
+
+  def isLocalReader: Boolean = {
+    if (partitionSpecs.exists(_.isInstanceOf[PartialMapperPartitionSpec])) {
+      assert(partitionSpecs.forall(_.isInstanceOf[PartialMapperPartitionSpec]))
+      true
+    } else {
+      false
+    }
+  }
+
+  private def shuffleStage = child match {
+    case stage: ShuffleQueryStageExec => Some(stage)
+    case _ => None
+  }
+
+  private lazy val partitionDataSizeMetrics = {
+    val maxSize = SQLMetrics.createSizeMetric(sparkContext, "maximum partition data size")
+    val minSize = SQLMetrics.createSizeMetric(sparkContext, "minimum partition data size")
+    val avgSize = SQLMetrics.createSizeMetric(sparkContext, "average partition data size")
+    val mapStats = shuffleStage.get.mapStats.bytesByPartitionId
+    val sizes = partitionSpecs.map {
```

Comment:
It is probably faster if you compute min/max/avg on the fly.

Reply:
it is, but probably the perf here doesn't matter too much.